vocode/streaming/agent/web_socket_user_implemented_agent.py

```diff
@@ -0,0 +1,145 @@
+import asyncio
+import json
+import logging
+from typing import Dict
+from vocode.streaming.transcriber.base_transcriber import Transcription
+from vocode.streaming.utils.worker import InterruptibleEvent
+import websockets
+from websockets.client import (
+    connect,
+    WebSocketClientProtocol,
+)
+
+from typing import Awaitable, Callable, Optional, cast
+from vocode.streaming.agent.base_agent import (
+    AgentInput,
+    AgentResponse,
+    AgentResponseMessage,
+    AgentResponseStop,
+    AgentResponseFillerAudio,
+    BaseAgent,
+)
+from vocode.streaming.models.message import BaseMessage
+from vocode.streaming.models.websocket_agent import (
+    WebSocketAgentMessage,
+    WebSocketAgentStopMessage,
+    WebSocketAgentTextMessage,
+    WebSocketUserImplementedAgentConfig,
+)
+
+NUM_RESTARTS = 5
+
+
+class WebSocketUserImplementedAgent(BaseAgent[WebSocketUserImplementedAgentConfig]):
+    input_queue: asyncio.Queue[InterruptibleEvent[AgentInput]]
+    output_queue: asyncio.Queue[InterruptibleEvent[AgentResponse]]
+
+    def __init__(
+        self,
+        agent_config: WebSocketUserImplementedAgentConfig,
+        logger: Optional[logging.Logger] = None,
+    ):
+        self.agent_config = agent_config
+        self.logger = logger or logging.getLogger(__name__)
+
+        self.has_ended = False
+        super().__init__(agent_config=agent_config, logger=logger)
+
+    def get_agent_config(self) -> WebSocketUserImplementedAgentConfig:
+        return self.agent_config
+
+    async def _run_loop(self) -> None:
+        restarts = 0
+        self.logger.info("Starting Socket Agent")
+        while not self.has_ended and restarts < NUM_RESTARTS:
+            await self._process()
+            restarts += 1
+            self.logger.debug(
+                "Socket Agent connection died, restarting, num_restarts: %s", restarts
+            )
+
+    def _handle_incoming_socket_message(self, message: WebSocketAgentMessage) -> None:
+        self.logger.info("Handling incoming message from Socket Agent: %s", message)
+
+        agent_response: AgentResponse
+
+        if isinstance(message, WebSocketAgentTextMessage):
+            agent_response = cast(AgentResponse, AgentResponseMessage(message=BaseMessage(text=message.data.text)))
```

Comment:
is this `cast` needed?

Reply:
Pydantic was showing an error without it. I agree it seems super strange to need to explicitly upcast an object like this.

Reply:
looks like it's because `AgentResponse` isn't actually the super-class (our bad)